oak_runtime/src/lib.rs

```diff
@@ -1138,9 +1163,12 @@ impl Runtime {
             )
         };
 
+        let node_infos = self.node_infos.read().unwrap();
+        let node_name = &node_infos.get(&node_id).expect("Invalid node_id").name;
```

Comment:
Does this not deadlock with the `self.node_infos().write()` on line 1152?  (I thought `RwLock` didn't allow upgrade, but I might be wrong).  

(Also, this is inducing an implied locking hierarchy with the `Mutex` around the `introspection_event_queue`, which is probably OK but needs a bit more thought / documentation.)

The obvious suggestion would be to encapsulate these two lines in a helper function, which ensures the lock is dropped when the function returns; however, that can't return a reference to the `String` in the `NodeInfo` and so would involve a bunch of building copies.

Reply:
Great catch! It does indeed deadlock. In this specific case it's sufficient to wrap this block in in its own scope. A helper function seems useful though, since I encountered a few deadlocks in other parts of the file if I didn't scope the `RwLock`.

Is there a better way to catch / debug this? In my case it didn't work up until I wrapped all instances of this in a helper function, which seemed pretty brute force. 

PS: Apologies for getting back on this so late, I was out for a few days.

Reply:
Yeah, at the moment I'd probably go for a `Runtime` method like:
```Rust
impl Runtime {
    def node_name(node_id: NodeId) -> String {
        let node_infos = self.node_infos.read().unwrap();
        node_infos.get(&node_id).expect("Invalid node_id").name.clone()
    }
}
```
and for now just live with the allocation every time &ndash; if it turns out to be significantly expensive later, we can revisit it.

(For example, we could have shenanigans where the non-`oak_debug` version of the method returns something different:
```Rust
impl Runtime {
    #[cfg(feature = "oak_debug")]
    def node_name(node_id: NodeId) -> String {
        let node_infos = self.node_infos.read().unwrap();
        node_infos.get(&node_id).expect("Invalid node_id").name.clone()
    }
    #[cfg(not(feature = "oak_debug"))]
    def node_name(node_id: NodeId) -> &'static str {
        "unavailable"
    }
}

```
given that logging will be turned off in that situation anyway).

For the more general problem, it does seem that Rust helps with the basics of multithreaded programming (by ensuring that shared state must always be synchronized), but doesn't have anything that helps with the more subtle (and more difficult) problems of locking hierarchies and inversions.  

ThreadSanitizer (which our CI does run) is the only tool I know of for catching those kind of things, albeit at runtime; otherwise it's a case of carefully looking over the design and the code to:
 - reduce the scope of all locks 
     - ideally by always encapsulating in scoped helpers
     - especially where the scope of one lock intersects with the scope of another
 - clearly document any lock orderings that are necessary.

I've done a couple of passes of that manual process for the Runtime in the past, but "constant vigilance" is hardly a software engineering process that scales over time &hellip;

Reply:
I agree with @daviddrysdale , a dedicated method would help avoiding the risk of long-lived locks, and also seems generally a good pattern to encapsulate the same logic that otherwise occurs quite a few times in this PR.

Reply:
Thanks for explaining @daviddrysdale! The context is really helpful. Having worked in a single threaded language before I was thankfully shielded from the complexity of locking hierarchies and inversions.

Reply:
Also added the helper method, def makes things easier.